File: clear_cache.php
<?php
/**
 * Cache clearing utility
 * 
 * Usage: php clear_cache.php
 * Or visit: https://yourdomain.com/clear_cache.php?confirm=1
 */

// Security: Require confirmation parameter in production
$confirm = $_GET['confirm'] ?? $_SERVER['argv'][1] ?? '0';

if ($confirm !== '1' && $confirm !== 'yes') {
    echo "Cache Clear Utility\n";
    echo "===================\n\n";
    echo "To clear all caches, run:\n";
    echo "  php clear_cache.php 1\n";
    echo "Or visit: " . (isset($_SERVER['HTTP_HOST']) ? "https://{$_SERVER['HTTP_HOST']}/clear_cache.php?confirm=1" : "this script with ?confirm=1") . "\n\n";
    exit(1);
}

echo "Clearing caches...\n\n";

// 1. Clear PHP opcache
if (function_exists('opcache_reset')) {
    if (opcache_reset()) {
        echo "✓ PHP opcache cleared\n";
    } else {
        echo "✗ PHP opcache reset failed (may not be enabled)\n";
    }
} else {
    echo "ℹ PHP opcache not available\n";
}

// 2. Clear APCu cache if available
if (function_exists('apcu_clear_cache')) {
    if (apcu_clear_cache()) {
        echo "✓ APCu cache cleared\n";
    } else {
        echo "✗ APCu cache clear failed\n";
    }
} else {
    echo "ℹ APCu not available\n";
}

// 3. Clear Python __pycache__ directories
$backendDir = __DIR__ . '/backend/src';
if (is_dir($backendDir)) {
    $cleared = false;
    $iterator = new RecursiveIteratorIterator(
        new RecursiveDirectoryIterator($backendDir, RecursiveDirectoryIterator::SKIP_DOTS),
        RecursiveIteratorIterator::CHILD_FIRST
    );
    
    foreach ($iterator as $path) {
        if ($path->isDir() && basename($path) === '__pycache__') {
            $files = glob($path->getPathname() . '/*.pyc');
            foreach ($files as $file) {
                @unlink($file);
            }
            if (@rmdir($path->getPathname())) {
                $cleared = true;
            }
        }
    }
    
    if ($cleared) {
        echo "✓ Python __pycache__ directories cleared\n";
    } else {
        echo "ℹ No Python __pycache__ directories found\n";
    }
} else {
    echo "ℹ Backend directory not found\n";
}

// 4. Clear session cache
session_start();
$_SESSION = [];
if (session_destroy()) {
    session_start(); // Restart session
    echo "✓ Session cache cleared\n";
} else {
    echo "ℹ Session cache clear skipped\n";
}

echo "\n✓ Cache clearing complete!\n";

File: backend/clear_cache.py
#!/usr/bin/env python3
"""
Cache clearing utility for Python
Clears all __pycache__ directories and .pyc files
"""

import os
import shutil
import sys
from pathlib import Path

def clear_pycache(directory=None):
    """Clear all __pycache__ directories and .pyc files"""
    if directory is None:
        directory = Path(__file__).parent
    
    directory = Path(directory)
    
    if not directory.exists():
        print(f"Directory not found: {directory}")
        return False
    
    cleared_count = 0
    
    # Find all __pycache__ directories
    for pycache_dir in directory.rglob('__pycache__'):
        if pycache_dir.is_dir():
            try:
                # Remove all .pyc files
                for pyc_file in pycache_dir.glob('*.pyc'):
                    pyc_file.unlink()
                    cleared_count += 1
                    print(f"Removed: {pyc_file}")
                
                # Remove the __pycache__ directory
                pycache_dir.rmdir()
                cleared_count += 1
                print(f"Removed: {pycache_dir}")
            except Exception as e:
                print(f"Error removing {pycache_dir}: {e}")
    
    # Find any stray .pyc files
    for pyc_file in directory.rglob('*.pyc'):
        try:
            pyc_file.unlink()
            cleared_count += 1
            print(f"Removed: {pyc_file}")
        except Exception as e:
            print(f"Error removing {pyc_file}: {e}")
    
    if cleared_count > 0:
        print(f"\n✓ Cleared {cleared_count} cache items")
        return True
    else:
        print("\nℹ No cache files found")
        return False

if __name__ == '__main__':
    target_dir = sys.argv[1] if len(sys.argv) > 1 else None
    success = clear_pycache(target_dir)
    sys.exit(0 if success else 1)

